blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,145 @@
+
+In this blog, we delve into the intricacies of Kubernetes’ `externalTrafficPolicy=Local` setting and explore strategies to gracefully handle pod shutdowns, rolling updates, and pod distribution. By following these best practices, you can enhance the resilience and reliability of your services while optimizing resource utilization across your AKS clusters.
+
+## `externalTrafficPolicy`: Cluster vs. Local
+
+**What does `externalTrafficPolicy` do?** In short, it influences how a cloud load balancer directs incoming traffic to Kubernetes nodes and how that traffic may hop between nodes to reach pods:
+
+- **Cluster Mode (default) (`externalTrafficPolicy=Cluster`)**:
+    All nodes in the cluster are included behind the Azure Load Balancer (ALB), regardless of whether they host any pods for the service. Incoming external traffic is distributed evenly across all nodes in the cluster. If a node receives traffic but doesn't have a local pod for the Service, Kubernetes (`kube-proxy`) will forward the request across the cluster network to a node that does. While this mode simplifies traffic distribution and reduces the risk of uneven load, it does not preserve the client’s original source IP and can potentially add latency and more failures modes due to extra hops between nodes. Alongside not preserving the source IP, using policy=cluster can also cause disruption to external traffic during events like node upgrades. Since the traffic is not directly sent to the node hosting a service pod, if any of the nodes are taken down, it could lead to 1/N (where N stands for node count) of the traffic to experience disruptions. So if you update your nodes weekly, this would disrupt some of your connections on a weekly basis.
+
+- **Local Mode (`externalTrafficPolicy=Local`)**:
+    In this mode, external traffic is directed only to nodes that host healthy pods for the specified service. The traffic is routed exclusively to the pods residing on the same node where it is received. This approach ensures that the client’s original source IP is preserved, which is critical for use cases involving security, logging, and analytics. However, it requires careful consideration of pod distribution across nodes to avoid uneven traffic loads.
+    If one node hosts multiple pods while another node hosts only one, the node with one pod will receive roughly the same share of incoming connections as the node with many pods.
+
+By understanding the operational differences between these two modes, you can make an informed decision based on your application’s requirements for traffic routing, source IP preservation, and load balancing.
+
+### Benefits of Local Mode
```

Comment:
This is a litle bit of a duplicatation of the previouse section with the exception of the healthCheckNodePort. Not a blocker but from a readers perspective, you could reqork to etierh combine these sections a bit more to highlight clearly the positioning of each traffic policy, or you could simplify the previous section where only hgih level descriptions are given and pros and cons are reserved for this next section? Just a thought. Again, not blocking.